### PR TITLE
Fix seeding of new generator for multi GPU

### LINKED ISSUE #3398: Randomization with seed not working
### System Info

```Shell
- `Accelerate` version: 0.32.1
- Platform: Linux-6.5.0-1022-aws-x86_64-with-glibc2.35
- Python version: 3.10.14
- Numpy version: 1.26.3
- PyTorch version (GPU?): 2.1.0+cu121 (True)
- PyTorch XPU available: False
- PyTorch NPU available: False
- PyTorch MLU available: False
- System RAM: 1121.80 GB
- GPU type: NVIDIA A100-SXM4-40GB
```

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [x] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [ ] My own task or dataset (give details below)

### Reproduction

Basically, what I am trying to do, is run a finetuning script on multi-GPU setting (Distributed Data Parallel). My setup looks like this : [https://github.com/huggingface/transformers/blob/main/examples/pytorch/language-modeling/run_mlm_no_trainer.py#L311](url).

Let me provide a toy example to exhibit what is the problem.

```python
import torch
from accelerate import Accelerator, DataLoaderConfiguration
from accelerate.utils import set_seed

set_seed(0)

accelerator = Accelerator()
dataset = range(10)
dataloader = torch.utils.data.DataLoader(dataset, batch_size=1, shuffle=True)
accelerator.print(type(dataloader))

dataloader = accelerator.prepare(dataloader)
accelerator.print(type(dataloader))

epochs=2

for epoch in range(epochs):
    accelerator.wait_for_everyone()
    accelerator.print(f"==============================================")
    accelerator.print(f"Epoch {epoch}")
    
    accelerator.wait_for_everyone()
    for idx, batch in enumerate(dataloader):
        if idx == 0:
            print(f'Batch {idx}: {batch}')
        
```

### Expected behavior

If i run this script with python (one process), setting the random seed ensures reproducibility. 
Changing the seed ensures that I have different shuffling  of the data.

But, if I run the script with accelerate launch (multi-process), the dataloader shuffles the data in the same way every time, regardless of the random seed. For example, by changing the random seed, the first batch  remains the exact same, and the data gets shuffled in the same way.
This is a bit strange to me, as the set_seed function (transformers' or accelerate's) achieves reproducibility in my one-process experiments. But I wanted to run some experiments on a distributed scenario (Multi-GPU), and I noticed this.

You can run the script that i provided with multiple seeds (or even no seed if you comment the set_seed line), and the fetched batches will be the exact same.
I also tried to include 

```python
dataloader_config=DataLoaderConfiguration(
        use_seedable_sampler=True,
    )
```
in the Accelerator initialization. by doing this, the random seeds worked for the first epoch (I mean that when I changed the seed, the data was fetched in different ways). BUT, another problem occured: the shuffling before each epoch did not work. So this was not a proper fix (?).

Please let me know if there is any proper way of fixing this, to ensure reproducible experiments with multiple seeds. I cannot find any sufficient answer  on this, it seems like a bug to me.

### PR DESCRIPTION
**Pull request by albertcthomas**
# What does this PR do?

Fixes #3398 

The initial state of the new generator was always the same, whatever the seed passed to `accelerate.utils.set_seed`. This PR makes sure that we get reproducible results but different numbers for different seeds passed to `set_seed`. This is already the case when multi GPU is disabled because generator is None (see https://github.com/pytorch/pytorch/blob/main/torch/utils/data/sampler.py#L170).
Before adding a test, I will wait for a first reviewer to confirm this is OK, I am not yet very familiar with accelerate.

Maybe the same should be done for https://github.com/huggingface/accelerate/blob/main/src/accelerate/data_loader.py#L1184 where the seed is set to 42 and thus will always produce the same number whatever the first global initial seed passed to `set_seed`

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

Based on git blame I would say @muellerzr :)

### CODE REVIEW

**Review by muellerzr — APPROVED** [review]
Nice catch, thanks!

**Comment by muellerzr** [comment]
Yes the same should be done there. Do you want to include it in this PR?

**Comment by albertcthomas** [comment]
> Yes the same should be done there. Do you want to include it in this PR?

Thanks for the quick check. Yes I will add it in this PR. I'll try to add a test as well.

**Comment by albertcthomas** [comment]
> Yes I will add it in this PR. I'll try to add a test as well.

Done.

**Comment by albertcthomas** [comment]
The added test fails on main, pass on this branch. Not sure this covers the other line I just changed in the code.

**Review by muellerzr — APPROVED** [review]
Thanks for adding a test!

**Comment by albertcthomas** [comment]
Thanks for the quick review!
